programs/util.c

```diff
@@ -0,0 +1,243 @@
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License along
+    with this program; if not, write to the Free Software Foundation, Inc.,
+    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+
+#if defined (__cplusplus)
+extern "C" {
+#endif
+
+
+/*-****************************************
+*  Dependencies
+******************************************/
+#include "util.h"   /* note : ensure that platform.h is included first ! */
+#include <stdio.h>  /* FILE*, perror */
+#include <errno.h>
+#include <assert.h>
+
+/*-****************************************
+*  count the number of cores
+******************************************/
+
+#if (defined(_WIN32) || defined(WIN32)) && defined(PSYSTEM_LOGICAL_PROCESSOR_INFORMATION)
+
+#include <windows.h>
+
+typedef BOOL(WINAPI* LPFN_GLPI)(PSYSTEM_LOGICAL_PROCESSOR_INFORMATION, PDWORD);
+
+DWORD CountSetBits(ULONG_PTR bitMask)
+{
+    DWORD LSHIFT = sizeof(ULONG_PTR)*8 - 1;
+    DWORD bitSetCount = 0;
+    ULONG_PTR bitTest = (ULONG_PTR)1 << LSHIFT;
+    DWORD i;
+
+    for (i = 0; i <= LSHIFT; ++i) {
+        bitSetCount += ((bitMask & bitTest)?1:0);
+        bitTest/=2;
+    }
+
+    return bitSetCount;
+}
+
+int UTIL_countCores(void)
+{
+    static int numCores = 0;
+    if (numCores != 0) return numCores;
+
+    {   LPFN_GLPI glpi;
+        BOOL done = FALSE;
+        PSYSTEM_LOGICAL_PROCESSOR_INFORMATION buffer = NULL;
+        PSYSTEM_LOGICAL_PROCESSOR_INFORMATION ptr = NULL;
+        DWORD returnLength = 0;
+        size_t byteOffset = 0;
+
+#if defined(_MSC_VER)
+/* Visual Studio does not like the following cast */
+#   pragma warning( disable : 4054 )  /* conversion from function ptr to data ptr */
+#   pragma warning( disable : 4055 )  /* conversion from data ptr to function ptr */
+#endif
+        HMODULE hModule = GetModuleHandle(TEXT("kernel32"));
+        if (hModule == NULL) {
+            goto failed;
+        }
+        glpi = (LPFN_GLPI)(void*)GetProcAddress(hModule,
+                                               "GetLogicalProcessorInformation");
+        if (glpi == NULL) {
+            goto failed;
+        }
+
+        while(!done) {
+            DWORD rc = glpi(buffer, &returnLength);
+            if (FALSE == rc) {
+                if (GetLastError() == ERROR_INSUFFICIENT_BUFFER) {
+                    if (buffer)
+                        free(buffer);
+                    buffer = (PSYSTEM_LOGICAL_PROCESSOR_INFORMATION)malloc(returnLength);
+
+                    if (buffer == NULL) {
+                        perror("lz4");
+                        exit(1);
+                    }
+                } else {
+                    /* some other error */
+                    goto failed;
```

Comment:
It may not happen.  But we can think about the following scenario : 
- `glpi()` returns `ERROR_INSUFFICIENT_BUFFER` at the first time.
- `buffer` is allocated, `while()` loop continues.
- `glpi()` fails at the second time. (for some reason)
- Here, potentially `buffer` will be leaked.